processor/tailsamplingprocessor/sampling/composite.go

```diff
@@ -0,0 +1,161 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package sampling
+
+import (
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+)
+
+type subpolicy struct {
+	// the subpolicy evaluator
+	evaluator PolicyEvaluator
+
+	// spans per second allocated to each subpolicy
+	allocatedSPS int64
+
+	// spans per second that each subpolicy sampled in this period
+	sampledSPS int64
+}
+
+// Composite evaluator and its internal data
+type Composite struct {
+	// the subpolicy evaluators
+	subpolicies []*subpolicy
+
+	// maximum total spans per second that must be sampled
+	maxTotalSPS int64
+
+	// current unix timestamp second
+	currentSecond int64
+
+	// The time provider (can be different from clock for testing purposes)
+	timeProvider TimeProvider
+
+	logger *zap.Logger
+}
+
+var _ PolicyEvaluator = (*Composite)(nil)
+
+// SubPolicyEvalParams defines the evaluator and max rate for a sub-policy
+type SubPolicyEvalParams struct {
+	Evaluator         PolicyEvaluator
+	MaxSpansPerSecond int64
+}
+
+// NewComposite creates a policy evaluator that samples all subpolicies.
+func NewComposite(
+	logger *zap.Logger,
+	maxTotalSpansPerSecond int64,
+	subPolicyParams []SubPolicyEvalParams,
+	timeProvider TimeProvider,
+) PolicyEvaluator {
+
+	subpolicies := []*subpolicy{}
+
+	for i := 0; i < len(subPolicyParams); i++ {
+		sub := &subpolicy{}
+		sub.evaluator = subPolicyParams[i].Evaluator
+		sub.allocatedSPS = subPolicyParams[i].MaxSpansPerSecond
+
+		// We are just starting, so there is no previous input, set it to 0
+		sub.sampledSPS = 0
+
+		subpolicies = append(subpolicies, sub)
+	}
+
+	return &Composite{
+		maxTotalSPS:  maxTotalSpansPerSecond,
+		subpolicies:  subpolicies,
+		timeProvider: timeProvider,
+		logger:       logger,
+	}
+}
+
+// OnLateArrivingSpans notifies the evaluator that the given list of spans arrived
+// after the sampling decision was already taken for the trace.
+// This gives the evaluator a chance to log any message/metrics and/or update any
+// related internal state.
+func (c *Composite) OnLateArrivingSpans(Decision, []*pdata.Span) error {
+	return nil
+}
+
+// Evaluate looks at the trace data and returns a corresponding SamplingDecision.
+func (c *Composite) Evaluate(traceID pdata.TraceID, trace *TraceData) (Decision, error) {
+	// func (c *Composite) Evaluate(traceID []byte, trace *TraceData) (Decision, error) {
+	// Rate limiting works by counting spans that are sampled during each 1 second
+	// time period. Until the total number of spans during a particular second
```

Comment:
I feel like this func could be simplified a lot by just using https://godoc.org/golang.org/x/time/rate .